<launch>

  <arg name="robot" default="firefly_sbx"/>
  <arg name="fixed_frame_id" default="global"/>
  <arg name="calibration_file"
    default="$(find orcvio)/config/camchain-imucam-dcist-erl.yaml"/>
  <arg name="visualize_rviz_arg" default="true" />

  <arg name="play_rosbag" default="true" />
  <arg name="republish" default="true" />

  <!-- path for rosbags  -->
  <!-- <arg name="rosbag_path" default="" /> -->
  <!-- <arg name="rosbag_path" default="/media/erl/disk1/dcist_vio/dcist_cde_bags/mit_outdoor_big_loop_2021-04-13-17-34-30.bag" /> -->
  <!-- <arg name="rosbag_path" default="/media/erl/disk1/dcist_vio/dcist_cde_bags/mit_outdoor_small_loop_2021-04-13-17-30-18.bag" /> -->
  <arg name="rosbag_path" default="/media/disk1/dcist_vio/dcist_cde_bags/kimera_big_loop_2021-05-04-16-40-46.bag" />
  
  <!-- open rviz  -->
  <!-- do not open rviz when running on jackal or uav  -->
  <group if="$(arg visualize_rviz_arg)" >
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find orcvio)/rviz/rviz_dcist_config.rviz"/>
  </group>

  <!-- Image Processor Nodelet  -->
  <include file="$(find orcvio)/launch/image_processor_dcist.launch">
    <arg name="robot" value="$(arg robot)"/>
    <arg name="calibration_file" value="$(arg calibration_file)"/>
  </include>

  <!-- OrcVio Nodelet  -->
  <group ns="$(arg robot)">
    <node pkg="nodelet" type="nodelet" name="vio"
      args='standalone orcvio/OrcVioNodelet'
      output="screen">

      <!-- Calibration parameters -->
      <rosparam command="load" file="$(arg calibration_file)"/>

      <param name="publish_tf" value="true"/>
      <param name="fixed_frame_id" value="$(arg fixed_frame_id)"/>
      <param name="child_frame_id" value="odom"/>
      <param name="max_cam_state_size" value="10"/>
      <param name="position_std_threshold" value="8.0"/>

      <param name="rotation_threshold" value="0.2618"/>
      <param name="translation_threshold" value="0.4"/>
      <param name="tracking_rate_threshold" value="0.5"/>

      <!-- Feature optimization config -->
      <param name="feature/config/translation_threshold" value="-1.0"/>

      <!-- These values should be standard deviation -->
      <param name="noise/gyro" value="0.005"/>
      <param name="noise/acc" value="0.05"/>
      <param name="noise/gyro_bias" value="0.001"/>
      <param name="noise/acc_bias" value="0.01"/>
      <param name="noise/feature" value="0.035"/>

      <param name="initial_state/velocity/x" value="0.0"/>
      <param name="initial_state/velocity/y" value="0.0"/>
      <param name="initial_state/velocity/z" value="0.0"/>

      <!-- These values should be covariance -->
      <param name="initial_covariance/velocity" value="0.01"/>
      <param name="initial_covariance/gyro_bias" value="0.01"/>
      <param name="initial_covariance/acc_bias" value="0.01"/>
      <param name="initial_covariance/extrinsic_rotation_cov" value="0"/>
      <param name="initial_covariance/extrinsic_translation_cov" value="0"/>

      <remap from="~imu" to="/acl_jackal/forward/imu"/>
      <remap from="~features" to="image_processor/features"/>

    </node>
  </group>

  <group if="$(arg republish)">
    <node name="republish1" type="republish" pkg="image_transport" output="screen" args="compressed in:=/acl_jackal/forward/infra1/image_rect_raw raw out:=/acl_jackal/forward/infra1/image_rect_raw" />
    <node name="republish2" type="republish" pkg="image_transport" output="screen" args="compressed in:=/acl_jackal/forward/infra2/image_rect_raw raw out:=/acl_jackal/forward/infra2/image_rect_raw" />
  </group>

  <!-- play the dataset -->
  <group if="$(arg play_rosbag)">
    <node pkg="tf" type="static_transform_publisher" name="tf" args="0.0 0.0 0.0 0.0 0.0 0.0 1.0 world global 1000"/>

    <node pkg="rosbag" type="play" name="rosbag" args="-s 80 $(arg rosbag_path)" required="true"/>
  </group>
  
</launch>